paseos/paseos.py

```diff
@@ -72,21 +69,47 @@ def save_status_log_csv(self, filename) -> None:
 
     def log_status(self):
         """Updates the status log."""
-        self._operations_monitor.log(self._local_actor, self.known_actor_names)
+        self._operations_monitor.log(self.local_actor, self.known_actor_names)
 
-    def advance_time(self, time_to_advance: float):
+    def advance_time(
+        self,
+        time_to_advance: float,
+        current_power_consumption_in_W: float,
+        constraint_function: types.FunctionType = None,
+    ):
         """Advances the simulation by a specified amount of time
 
         Args:
             time_to_advance (float): Time to advance in seconds.
+            current_power_consumption_in_W (float): Current power consumed per second in Watt.
+            constraint_function(FunctionType): Constraint function which will be evaluated
+            every cfg.sim.activity_timestep seconds. Aborts the advancement if False.
+
         """
+        assert time_to_advance > 0, "Time to advance has to be positive."
+        assert (
+            current_power_consumption_in_W >= 0
+        ), "Power consumption cannot be negative."
+
         logger.debug("Advancing time by " + str(time_to_advance) + " s.")
         target_time = self._state.time + time_to_advance
         dt = self._cfg.sim.dt
 
+        time_since_constraint_check = float("inf")
+
         # Perform timesteps until target_time - dt reached,
         # then final smaller or equal timestep to reach target_time
         while self._state.time < target_time:
+            if (
+                constraint_function is not None
+                and time_since_constraint_check > self._cfg.sim.activity_timestep
```

Comment:
what if we want to have both activities and an event? Maybe we should use another time_duration than activity_timestep?

Reply:
this is currently not really supported because then you have both an asynchronous and synchronous call of this function happening. I don't really see any way of supporting this without risking threading hell 🤔 

We should, however, throw errors to avoid the users doing this by accident, I guess? Will add

Reply:
I added a semaphore to the advance_time function now to actively disallow this and warn the user about thread-safety.

https://github.com/aidotse/PASEOS/pull/90/commits/1dbf77b4b9292766b354590c4fb58e72fd2bd408

Reply:
Mm, I guess one (ugly) way to support both would be to juggle the constraint function between the activity processor and the paseos instance. If an activity is running, the activity processor makes sure the constraint is passed. Perhaps not worth the effort...

Reply:
I see. It would be nice to have two event loops, one for handling these kinds of constraints and one for handling activities. I think it is possible to have separate event loops on different threads (https://stackoverflow.com/questions/31623194/asyncio-two-loops-for-different-i-o-tasks#:~:text=What%20I%20ended,worked%20for%20me.). But yea, this goes into the threading abyss and may not be worthwhile